### PR TITLE
update/api-core-tests guide for writing tests

### PR DESCRIPTION
**Pull request by nigeljamesstevenson**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:
Update api-core-tests documentation

Closes 451-gh-woocommerce/woocommerce-quality.

- [x] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. This is a documentation update and does not require testing

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: I noticed we're removing the reference to Playwright's debugging, does the debugging provided by Playwright not work well for API tests?

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Maybe here we can consider including some "good examples" (1-2) so folks have a good starting point when jumping in. We could include a test and how it uses the equivalent data file.

Minor wording change suggestions:

```suggestion
When writing new tests, a good source on how to get started is to reference the [existing tests](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/api-core-tests/tests). Data that is required for the tests should be located in an equivalent file in the [data](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/api-core-tests/data) folder.
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Some minor wording suggestions:

```suggestion
The [Playwright documentation](https://playwright.dev/docs/intro) is a good source for finding out more details on the various methods used in the tests, including what is available to you when you write new tests. The [API testing](https://playwright.dev/docs/test-api-testing) section has a good example on how the [playwright.config.js](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/api-core-tests/playwright.config.js) is used (see [Configuration](https://playwright.dev/docs/test-api-testing#configuration) section) and also how [Setup and Teardown](https://playwright.dev/docs/test-api-testing#setup-and-teardown) works. 
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Minor formatting suggestion:

```suggestion
Each test file requires the `@playwright/test` module to be imported as follows:
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Minor formatting:

```suggestion
This allows you to further subgroup tests. When viewing the tests results locally, each test describe 'level' will be separated by `>` in the console as below: 
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Minor addition for clarity:

```suggestion
You may need test data setup prior to the execution of your tests and removed after the execution of your tests. This can be achieved with any of the following methods, depending on the needs of the test:
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Some suggestions for this list:

```markdown
1. Create `test.js` file inside the tests directory
2. Import `@playwright/test` module
3. Group tests with `test.describe()` methods
4. Add tests with `test()` methods
5. Separate data where required into files in the `data` directory 
6. After writing your tests, ensure all tests pass successfully with `pnpm test:api-pw`
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Minor typo fixes:

```suggestion
If you have made updates to functionality that breaks the tests then the tests should be updated accordingly. Similarly, if there is new functionality added then new tests should be added.
```

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/README.md:257]: Love all these examples!

**Review by zhongruige — COMMENTED** [review]
Nice work on this @nigeljamesstevenson! Love these additions to the README 🙌 

Added a few comments around some wording changes, formatting, and a few typos. Let me know if you have any questions on any of it.

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md]: Great question - The debugging provided by Playwright uses Playwright Inspector (a GUI tool) which operates only when there are browser interactions so it isn't applicable for API tests.
e.g if you were debugging one of the e2e tests you would make a call such as `npx playwright test example --debug` or pnpm equivalent which would fire up the inspector 
This wouldn't have any effect with the API tests. With that being said, I will remove the line mentioning `pnpm test:api-pw --debug (debug)` currently in the guide.

I also did a spot of googling and noticed a [feature request](https://github.com/microsoft/playwright/issues/19177) to to make debugging easier for API tests (added last week) - I have subscribed and if that comes to the fore then I will update accordingly

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md]: > Maybe here we can consider including some "good examples" (1-2) so folks have a good starting point when jumping in. We could include a test and how it uses the equivalent data file.
> 
Great suggestion - I'll get this in

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md:133]: Should we briefly explain in this section that a test contains 3 different stages (`preconditions, action and validation`) and in concrete, in the case of API testing, we could extrapolate those sections to `data creation, request sending and response validation`?

This way we wouldn't assume that every person reading this section has knowledge of API testing or even general testing.

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md]: I think we could extend a little bit this 2nd point with a technical detail to be a bit redundant on what to check in the JSON response. We could say something like:
```
2. Verify response payload. Check the JSON body is valid and field names, types, and values are correct (i.e. check the response payload schema is implemented according to the specification) — including error responses.
```
Note: I took the liberty of changing some words in the example above to make the sentence more readable and easier to understand (at least for me, as a non-native English speaker). Feel free to discard this wording change if you don't find it relevant!

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md]: There is a typo in the `applicable` word.

I just tested ChatGPT with this and it spotted the typo 😄 
https://d.pr/i/fZSOJh

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md:177]: Should we maybe explain that with negative testing we intent to validate that we get error responses when expected as per specification and the error status code and message are correct as per documentation?

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md]: I think "You may need test data setup prior to the execution of your tests and `make sure it is removed` after the execution of your tests" is more readable for me as a non-native English speaker.

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Would it be worth clarifying that the order execution is this or perhaps it would be too redundant?:
1. test.beforeAll()
2. test.beforeEach()
3. test.afterEach()
4. test.afterAll()

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (alopezari) [inline plugins/woocommerce/tests/api-core-tests/README.md]: Should the `Variable` word be in lower-case?

**Review by alopezari — COMMENTED** [review]
Thank you very much for creating this awesome guide @nigeljamesstevenson! I added some minor comments, but it looks great.

I think an amazing way to test that the guide works would be to ask some developers to read it (since they would be the consumers of this guide) and check if they would feel comfortable to start contributing to the API tests afterwards (not expecting them to do so yet, just to check if they would feel ready for it). Do you think it would be interesting trying this? cc @zhongruige

**Comment by nigeljamesstevenson** [comment]
> Thank you very much for creating this awesome guide @nigeljamesstevenson! I added some minor comments, but it looks great.
> 
> I think an amazing way to test that the guide works would be to ask some developers to read it (since they would be the consumers of this guide) and check if they would feel comfortable to start contributing to the API tests afterwards (not expecting them to do so yet, just to check if they would feel ready for it). Do you think it would be interesting trying this? cc @zhongruige

Thanks for all the comments @alopezari (and always great to hear from @zhongruige , of course 😊)
Alex - appreciate the viewpoint of a non-native English speaker!

I will make the appropriate updates today and poke you both once completed to give it another once over. 👍

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md:133]: I love this idea and will add a new section that references each of the stages, along with examples in the guide...

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md]: I think it is worth re-ordering for emphasis

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [reply plugins/woocommerce/tests/api-core-tests/README.md]: > I also did a spot of googling and noticed a https://github.com/microsoft/playwright/issues/19177 to to make debugging easier for API tests

Nice! That would be cool to have that feature added in.

Debating if it's worth noting that the Playwright debugger won't work for the API tests, so for now it's easier/best to use `console.log()` statements to debug? Additionally, I think we could recommend using a plugin to see the API requests within WordPress, something like [this plugin](https://wordpress.org/plugins/wp-rest-api-log/)?

**Comment by alopezari** [comment]
Thank you very much for adding these changes so quickly @nigeljamesstevenson!

I've resolved all my comments and everything is ready from my side. I think we can approve this PR once @zhongruige's comments are resolved and nothing new arises 👍

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md]: Fully agree with this - the comment should be updated for clarity and informational purposes - good shout 👍

Also - great suggestion on the plugin - I'll take a peek and see how it looks 🙌🏻

**Review by zhongruige — DISMISSED** [review]
LGTM! Thanks for the great work on this @nigeljamesstevenson!

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (zhongruige) [reply plugins/woocommerce/tests/api-core-tests/README.md]: Thank you! The additions look good to me, thanks for adding this in!

**Comment by nigeljamesstevenson** [comment]
Added @rodelgc in as a reviewer to make sure none of his lovely guide is impacted after resolving some merge conflicts... 🙏

File: plugins/woocommerce/tests/api-core-tests/README.md
Comment (rodelgc) [inline plugins/woocommerce/tests/api-core-tests/README.md:438]: Good call to suggest this plugin! Very helpful indeed when debugging.

Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/README.md:438]: Glory goes to @zhongruige on this suggestion! I had a play about with the plugin and it looked great.

**Comment by rodelgc** [comment]
@nigeljamesstevenson LGTM! I could verify that none of my changes were omitted during the merge conflict resolution.

I wonder though if we need a section, even just a short one, to explain how the authentication of the requests is being set up? I'm just thinking of a scenario where a reader might go through the "Creating test structure" or "Writing the test" without being aware that they need to set up the authentication in the Playwright config file first.

**Comment by nigeljamesstevenson** [comment]
> @nigeljamesstevenson LGTM! I could verify that none of my changes were omitted during the merge conflict resolution.
> 
> I wonder though if we need a section, even just a short one, to explain how the authentication of the requests is being set up? I'm just thinking of a scenario where a reader might go through the "Creating test structure" or "Writing the test" without being aware that they need to set up the authentication in the Playwright config file first.

This is a great suggestion - I'll add something in and ping out for a look see once completed 👀

**Comment by nigeljamesstevenson** [comment]
@rodelgc I added a bullet into the quick start guide regarding authentication and referenced the appropriate section - hopefully this makes things clear. 🙏

If there is anything else, let me know - otherwise feel free to hit that approve button 😉👍

@alopezari and @zhongruige

**Review by zhongruige — APPROVED** [review]
LGTM! Thank you for these great documentation updates @nigeljamesstevenson!

**Comment by nigeljamesstevenson** [comment]
Hi folks, I am seeing that Merging is blocked on this PR - any ideas? Thanks 😊
@zhongruige @rodelgc @alopezari

**Pull request merged**
